walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -35,106 +40,263 @@ public int compare(Event a, Event b) {
-    for (Event event : events) {
-      // event can be merged with current time range
-      if (event.getWhen().start() <= end) {
-        end = Math.max(end, event.getWhen().end());
-        continue;
+
+    /**
+     * Returns all meeting times that allow the most optional attendees to attend. These times must
+     * also fall in a time satisfying all mandatory attendees and must be at least minTime long.
+     * Utilizes two pointers: one to mandatoryAttendeesMeetingTimes, one to changeLog.
+     *
+     * @param mandatoryAttendeesMeetingTimes all meeting times satisfying mandatory attendees
+     * @param changeLog a change log of the number of available optional attendees over time
+     * @param minTime the minimum time a meeting must last for
+     */
+    private Collection<TimeRange> getOptimalMeetingTimes(
+        ArrayList<TimeRange> mandatoryAttendeesMeetingTimes,
+        TreeMap<Integer, Integer> changes,
+        long minTime) {
+
+      ArrayList<TimeRange> optimalMeetingTimes = new ArrayList<TimeRange>();
+      ArrayList<Map.Entry<Integer, Integer>> changeLog =
+          new ArrayList<Map.Entry<Integer, Integer>>(changes.entrySet());
+
+      // changeLogIndexis a pointer in changeLog,  mandatoryAttendeesMeetingTimesIndex is a pointer
+      // in mandatoryAttendeesMeetingTimes.
+      int mandatoryAttendeesMeetingTimesIndex = 0,
+          currAttendance = 0,
+          bestAttendance = 0,
+          prevTime = 0;
+      for (int changeLogIndex = 0; changeLogIndex < changeLog.size(); ++changeLogIndex) {
+        // Need to back up  mandatoryAttendeesMeetingTimesIndex in case we missed a time range in
+        // mandatoryAttendeesMeetingTimes.
+        mandatoryAttendeesMeetingTimesIndex = Math.max(0, mandatoryAttendeesMeetingTimesIndex - 1);
+
+        // Compares time range from previous time in changeLog to current time in changeLog
+        // mandatoryAttendeesMeetingTimes that overlap with this time range.
+        for (;
+            mandatoryAttendeesMeetingTimesIndex < mandatoryAttendeesMeetingTimes.size()
+                && mandatoryAttendeesMeetingTimes.get(mandatoryAttendeesMeetingTimesIndex).start()
+                    < (Integer) changeLog.get(changeLogIndex).getKey();
+            mandatoryAttendeesMeetingTimesIndex++) {
+          bestAttendance =
+              updateOptimalTimes(
+                  TimeRange.fromStartEnd(
+                      Math.max(
+                          mandatoryAttendeesMeetingTimes
+                              .get(mandatoryAttendeesMeetingTimesIndex)
+                              .start(),
+                          prevTime),
+                      Math.min(
+                          mandatoryAttendeesMeetingTimes
+                              .get(mandatoryAttendeesMeetingTimesIndex)
+                              .end(),
+                          (Integer) changeLog.get(changeLogIndex).getKey()),
+                      false),
+                  bestAttendance,
```

Comment:
These should probably all be member variables (i assume you're writing to them... ugh).